### DISCUSSION TITLE
Is there any mocking library available to plug with protractor/serenity setup for Non-Angular apps?

### DISCUSSION THREAD

**Discussion by abhinaba-ghosh** [serenity-js/serenity-js#1646] [How to... ?]
Hi Jan,

I am having an non angular app under test. So, most of the time I am using browser.waitForAngularEnabled(false). In this circumstance, most of the general mocking library does not work with protractor (ex: protractor-http-mock etc) as, for non angular apps protractor can not track the http calls. 

I found out a library ng-apimock which can be useful but for already deployed app, it seems bit hard to proceed on. Refer my ticket here: https://github.com/mdasberg/ng-apimock/issues/84

So, have you encountered such issues in past? Can you suggest me a suitable mocking library or anything so that i can get start with.

**Comment by jan-molak** [comment]
Hey @abhinaba1080 - good question! The answer depends on the type of the framework your app uses and the affordances it provides. I.e. does the framework allow you to inject [test doubles](https://janmolak.com/mocking-javascript-a-little-26efb5f7f52a) in an elegant way like Angular does? Or would you have to mock things at a lower level, for example by using an [`xhr-mock`](https://www.npmjs.com/package/xhr-mock)?

One way or another, you'll probably be looking at loading some external mocking script using the [`ExecuteScript.from`](https://serenity-js.org/modules/protractor/class/src/screenplay/interactions/execute-script/ExecuteScript.ts~ExecuteScript.html#static-method-from), and then configuring the mocks using [`ExecuteScript.sync`](https://serenity-js.org/modules/protractor/class/src/screenplay/interactions/execute-script/ExecuteScript.ts~ExecuteScript.html#static-method-sync) or [`ExecuteScript.async`](https://serenity-js.org/modules/protractor/class/src/screenplay/interactions/execute-script/ExecuteScript.ts~ExecuteScript.html#static-method-async). 

For example:

```typescript
import { ExecuteScript } from '@serenity-js/protractor'l

actor.attemptsTo(
    ExecuteScript.from('https://unpkg.com/xhr-mock/dist/xhr-mock.js')
    ExecuteScript.sync(`
        window.XHRMock.post('/api/user', {
            status: 201,
            reason: 'Created',
            body: '{"data":{"id":"abc-123"}}'
        });
    `),
    DoSomethingThatTriggersTheAPICall(),
);
``` 

Does this help?

**Comment by abhinaba-ghosh** [comment]
Thank @jan-molak .

1. What does ExecuteScript.sync do? what is the alternate protractor browser.executeScript command?

2. i have tried the same by simply executing chrome console, it seems nothing happens. if time permits, can try a basic scenario in your chrome console and let me know the outcome.

**Comment by jan-molak** [comment]
All the `ExecuteScript` interactions inject JavaScript code into the browser used in the test.
You can use this feature to interact with your application at a low, programmatic level (a.k.a. do things that the user would not have normally done).

It's a rather advanced technique and is basically the same as if you typed JavaScript code into the Chrome console during the test. 

To learn more about those, check out the [tests demonstrating the behaviour](https://github.com/serenity-js/serenity-js/tree/master/packages/protractor/spec/screenplay/interactions/execute-script) of those interactions.
For example, here's what [`ExecuteScript.sync`](https://github.com/serenity-js/serenity-js/blob/master/packages/protractor/spec/screenplay/interactions/execute-script/ExecuteSynchronousScript.spec.ts) does, and here's the [implementation of those interactions](https://github.com/serenity-js/serenity-js/tree/master/packages/protractor/src/screenplay/interactions/execute-script) - that's where you can see the corresponding `protractor` methods.

If you wanted to execute my example in the console, you'd have to do this as follows:
```
document.write("<script src='https://unpkg.com/xhr-mock/dist/xhr-mock.js'></script>");
```
Followed by:
```
window.XHRMock.post('/api/user', {
            status: 201,
            reason: 'Created',
            body: '{"data":{"id":"abc-123"}}'
        });
```

---
_To people who read this post in the future - a word of warning - interacting with the system under test at a programmatic level rather than through an "official" interface (UI, REST API, etc.) couples the tests with how the system works under the hood. This can make your tests susceptible to changes introduced in the system so make sure everybody on your team knows that you're doing it and why._

**Comment by abhinaba-ghosh** [comment]
Hi @jan-molak 

Do we have in-built mocking utility inside the API library of Serenity 2.0?

Any documentation will help as I have just started to migrate the API testing framework from Serenity 1.x to 2.0.

**Comment by jan-molak** [comment]
Hey @abhinaba1080 - there are two ways you can do it.

### Axios Mock

The ability to `CallAnApi` uses `axios` under the hood, which means you can mock the responses using something like the [`axios-mock-adapter`](https://github.com/ctimmerm/axios-mock-adapter).

For example:

```typescript
import { engage, Actor, Cast } from '@serenity-js/core';
import { CallAnApi } from '@serenity-js/rest';
import axios from 'axios';
import MockAdapter from 'axios-mock-adapter';

class APIActors implements Cast {
     prepare(actor: Actor) {
        const axiosInstance = axios.create({
          // .. options, see https://github.com/axios/axios#creating-an-instance
        });
        const mock = new MockAdapter(axios);  // see https://github.com/ctimmerm/axios-mock-adapter

        mock.onGet('/users', { params: { searchText: 'John' } }).reply(200, {
            users: [
                { id: 1, name: 'John Smith' }
            ]
        });

        return actor.whoCan(CallAnApi.using(axiosInstance));
    }
}
```
Engage the in the before hook of your chosen test runner, i.e. Cucumber:
```typescript
Before(() => engage(new APIActors()));
```
Then in the test:
```typescript
When('......', () => actorInTheSpotlight().attemptsTo(
    Send.a(GetRequest.to('/users')),     // this will call the mock
));

Then('...', () => actorInTheSpotlight().attemptsTo(
    Ensure.that(LastResponse.status(), equals(200))
    Ensure.that(LastResponse.body(), equals( {
        users: [
            { id: 1, name: 'John Smith' }
        ]
    }),
));
```

However, what the test is doing is effectively just calling the mock.
This might be useful to prototype the test when the API is not ready yet or to test-drive the API design.

### Local server

Another way to do it is to use a fake API server via [`@serenity-js/local-server`](https://serenity-js.org/modules/local-server/).

For example:
```typescript
import { actorCalled } from '@serenity-js/core';
import {
    ManageALocalServer, StartLocalTestServer, StopLocalTestServer
} from '@serenity-js/local-server'
import { CallAnApi, GetRequest, Send } from '@serenity-js/rest';
import { Ensure, equals } from '@serenity-js/assertions';
import axios from 'axios';

const actor = actorCalled('Apisit').whoCan(
    ManageALocalServer.using(/* requestListener */),
    CallAnApi.using(axios.create()),
);

actor.attemptsTo(
    StartLocalTestServer.onRandomPort(),
    Send.a(GetRequest.to(LocalServer.url())),
    Ensure.that(LastResponse.status(), equals(200)),
    Ensure.that(LastResponse.body(), equals('Hello!')),
    StopLocalTestServer.ifRunning(),
);
```

The `requestListener` you see in the code sample above can be any `restify`, `koa`, `hapi`, `express`, or other Node.js server. This means that you can either use one of the many [dedicated node mock servers](https://www.npmjs.com/search?q=mock%20server), or write your own.

The nice thing about using this approach is that the `@serenity-js/local-server` starts an actual node.js HTTP or HTTPS server, which you can connect your UI to if you needed.

---

Let me know which approach you decided to take!
Jan

**Comment by abhinaba-ghosh** [comment]
Hi @jan-molak , 
I was trying stuff out with Axios-mock-adapter. I am getting below error:

```
Given user is logged in application with username as "admin"
    oar_api_admin wants authenticate user= admin
      ✗ oar_api_admin sends a POST request to '/api/login' (5ms)
        TypeError: Cannot convert undefined or null to object
  ⇢ When user sends GET request to the resource "users"
  ⇢ Then user receives success response from the API resource

✗ Execution failed with error (56ms)

  TypeError: Cannot convert undefined or null to object
      at Function.keys (<anonymous>)
      at mapToString (C:\Dev Room\Projects\ArcherAPIBDDTests\node_modules\@serenity-js\serenity-bdd\src\stage\crew\serenity-bdd-reporter\reports\SceneReport.ts:269:27)
      at C:\Dev Room\Projects\ArcherAPIBDDTests\node_modules\@serenity-js\serenity-bdd\src\stage\crew\serenity-bdd-reporter\reports\SceneReport.ts:281:34
      at SceneReport.withMutated (C:\Dev Room\Projects\ArcherAPIBDDTests\node_modules\@serenity-js\serenity-bdd\src\stage\crew\serenity-bdd-reporter\reports\SceneReport.ts:351:9)
      at SceneReport.httpRequestCaptured (C:\Dev Room\Projects\ArcherAPIBDDTests\node_modules\@serenity-js\serenity-bdd\src\stage\crew\serenity-bdd-reporter\reports\SceneReport.ts:272:21)
      at MatchesObjectsWithCommonPrototype.transformation (C:\Dev Room\Projects\ArcherAPIBDDTests\node_modules\@serenity-js\serenity-bdd\src\stage\crew\serenity-bdd-reporter\strategies\SceneReportingStrategy.ts:40:57)
      at MatchesObjectsWithCommonPrototype.MatcherRule.execute (C:\Dev Room\Projects\ArcherAPIBDDTests\node_modules\tiny-types\src\pattern-matching\rules\MatcherRule.ts:13:21)
      at ObjectMatcher.PatternMatcher.else (C:\Dev Room\Projects\ArcherAPIBDDTests\node_modules\tiny-types\src\pattern-matching\PatternMatcher.ts:17:71)
      at MatchesObjectsWithCommonPrototype.transformation (C:\Dev Room\Projects\ArcherAPIBDDTests\node_modules\@serenity-js\serenity-bdd\src\stage\crew\serenity-bdd-reporter\strategies\SceneReportingStrategy.ts:49:22)
      at MatchesObjectsWithCommonPrototype.MatcherRule.execute (C:\Dev Room\Projects\ArcherAPIBDDTests\node_modules\tiny-types\src\pattern-matching\rules\MatcherRule.ts:13:21)
      at ObjectMatcher.PatternMatcher.else (C:\Dev Room\Projects\ArcherAPIBDDTests\node_modules\tiny-types\src\pattern-matching\PatternMatcher.ts:17:71)
      at SingleSceneReportingStrategy.SceneReportingStrategy.handle (C:\Dev Room\Projects\ArcherAPIBDDTests\node_modules\@serenity-js\serenity-bdd\src\stage\crew\serenity-bdd-reporter\strategies\SceneReportingStrategy.ts:53:18)
      at MatchesAnything.transformation (C:\Dev Room\Projects\ArcherAPIBDDTests\node_modules\@serenity-js\serenity-bdd\src\stage\crew\serenity-bdd-reporter\strategies\SingleSceneReportingStrategy.ts:26:36)
      at MatchesAnything.MatcherRule.execute (C:\Dev Room\Projects\ArcherAPIBDDTests\node_modules\tiny-types\src\pattern-matching\rules\MatcherRule.ts:13:21)
      at ObjectMatcher.PatternMatcher.else (C:\Dev Room\Projects\ArcherAPIBDDTests\node_modules\tiny-types\src\pattern-matching\PatternMatcher.ts:17:71)
      at SingleSceneReportingStrategy.handle (C:\Dev Room\Projects\ArcherAPIBDDTests\node_modules\@serenity-js\serenity-bdd\src\stage\crew\serenity-bdd-reporter\strategies\SingleSceneReportingStrategy.ts:26:18)
```

My Actors are implemented like below:

```
export class Actors implements Cast {
  prepare(actor: Actor): Actor {
    const mock: MockAdapter = new MockAdapter(axiosClient);

    mock.onPost('/api/login').reply(200);

    return actor.whoCan(CallAnApi.using(axiosClient));
  }
}

```
The Axios Client looks like:
```
const agent = tunnel.httpsOverHttp({
   proxy: {
     host: process.env.API_PROXY_HOST,
     port: Number(process.env.API_PROXY_PORT)
   }
 });

const axiosClient: AxiosInstance = axios.create({
  baseURL: process.env.URL,
  httpsAgent: agent,
  proxy: false
});
```
I think, ```const mock: MockAdapter = new MockAdapter(axiosClient);``` - this initialization is throwing error somehow. 
Can you guide me here?

**Comment by abhinaba-ghosh** [comment]
Hey just analyzed that if I remove SerenityBDDReporter from the crew, everything works fine. Is there any modification needed for SerenityBDDReporter?
